harvester/harvest.py

```diff
@@ -507,12 +531,19 @@ def sync(self) -> None:
             if self.action == "update":
                 self.update_record()
         except Exception as e:
-            self.status = "error"
-            raise SynchronizeException(
-                f"failed to {self.action} for {self.identifier} :: {repr(e)}",
-                self.harvest_source.job_id,
-                self.harvest_source.internal_records_lookup_table[self.identifier],
-            )
+            if iter_count == 0 and CKAN_ERRORS["URL_IN_USE"] in repr(e):
+                self.ckanified_metadata["name"] = add_uuid_to_package_name(
+                    self.ckanified_metadata["name"]
+                )
+                self.ckan_name = self.ckanified_metadata["name"]
+                return self.sync(iter_count=1)
+            else:
+                self.status = "error"
+                raise SynchronizeException(
+                    f"failed to {self.action} for {self.identifier} :: {repr(e)}",
+                    self.harvest_source.job_id,
+                    self.harvest_source.internal_records_lookup_table[self.identifier],
+                )
```

Comment:
Maybe this is nitpicky, but I disagree that this should be handled as an exception. I believe this is a CKAN problem, and should be integrated into the CKAN utils library. It should retry with this function if the exception occurs, not pass back to the harvester system and then go back to CKAN.

Reply:
Do you mean adding this 5-char UUID to every dataset, or checking for naming conflicts prior to the POST, or some third thing?

>  I believe this is a CKAN problem, and should be integrated into the CKAN utils library. It should retry with this function if the exception occurs, not pass back to the harvester system and then go back to CKAN.

I opted for the recursion since I didn't want to replicate the network call.

Reply:
is the ckan exception identifiable to where we could handle it like this? or do you have to parse the message it sends to figure out what happened? I know there's redundancy in ckanifying the record again but this seems simpler maybe? i feel like the likelihood of an infinite loop is very small.
```python
#...
try:
    if self.action == "create":
        self.create_record()
    if self.action == "update":
        self.update_record()
except CKANValidation as e:
    self.ckanified_metadata["name"] = add_uuid_to_package_name(
        self.ckanified_metadata["name"]
    )
    self.sync()
# every other exception?...
except Exception as e:
    self.status = "error"
    raise SynchronizeException(
        f"failed to {self.action} for {self.identifier} :: {repr(e)}",
        self.harvest_source.job_id,
        self.harvest_source.internal_records_lookup_table[self.identifier],
    )
#...
```

Reply:
oh wait that wouldn't work

Reply:
I guess what I'm saying is that the create and update functions are just 1-liners, and it makes more sense to handle this logic there.
Also looking from a user perspective: I wouldn't expect a title change to change the name, that changes the URL for the user. Once the item is created, the URL should stay static (it shouldn't re-munge the name). So error catching that and re-trying when updating doesn't make sense to me. So this should be error handled and re-tried recursively (limited to 1 extra try) at the [create_record](https://github.com/GSA/datagov-harvester/pull/90/files#diff-598762acc9262886a623d9784cd9a5666592e736c2ae4df37c7d921673054e6eR476) level, and not this level.

Reply:
i included `update_record` because I assumed a title change could happen but if that's uncommon i see the reason to apply the re-try logic just to `create_record`

Reply:
This is great feedback. I moved the retry loop to the `create_record` method and now that we have `ckan_name` in place the name won't change on update even if the title does. Going to write a new test to confirm my nested try/except bubbles up to the outer one and then will push my changes.